### ISSUE TITLE
File substitution

### ISSUE THREAD

**Issue by bonzini** [ninja-build/ninja#1447]
For platforms (such as OS X) where there is no support for response files, it would be a useful addition to Ninja if command line arguments could come from a file. Here is one real world example taken from QEMU. The build system here would like to add several "-u" options to the linker command line, but the full list is not known until build time. With this extension, Ninja this could handle it like this:

     rule undefsym:
        command = ./build-undefsym.py $out $in
     rule link: 
        command = $ld -o $out $in $@UNDEF

    build somelibrary.undef: undefsym a.o b.o
    build myprogram: link p.o q.o | somelibrary.undef
        UNDEF = somelibrary.undef

In other words, `$@UNDEF` treats the variable as a file name and puts each line of the file into a separate command line argument.

(See [here](https://github.com/mesonbuild/meson/pull/3683) for a sample Meson pull request that produces rules similar to these, except that it uses response files to include the `.undef` file on the command line).

I would like some remarks on the feature and the syntax; I'm happy to implement it myself if the idea is accepted.

**Comment by atetubou** [comment]
Wrapper script for $ld is not sufficient?

**Comment by bonzini** [comment]
It sort of works, but I don't think it's satisfactory and it's certainly ugly. The problem is that the wrapper script would have to understand response files in order to avoid hitting command-line limits, and produce a second response file with the file contents included. But that means the wrapper script would also have to know the value of rspfile_content, for example.

Ninja can do it directly when writing the response file. In fact, it could special case `$@UNDEF` in rspfile_contents and just replace `$@UNDEF` with `@somelibrary.undef`, since nested response files work. Then `somelibrary.undef` would be read in by the linker and Ninja would not waste time writing a larger response file (IIRC writing the response file is serializing).

**Comment by bonzini** [comment]
Is the silence a sign that I should go on and do the patch? :)

**Comment by bonzini** [comment]
-filelist can only contain file names rather than arbitrary command line options.

**Comment by bonzini** [comment]
I did a first prototype and it's admittedly not pretty. [Here](https://github.com/ninja-build/ninja/compare/master...bonzini:file-subst?expand=1) you can see what this entails; no tests or docs changes yet, but at least you can give a quick check whether it is too intrusive or acceptable.

**Comment by evmar** [comment]
I don't quite follow the bug.  How is this different than something like

```
command = foo $(cat myfile)
```

?

**Comment by bonzini** [comment]
The handling of shell quoting, for example. However that doesn't matter for my use case, so feel free to close if the patches are just too ugly.

**Comment by bonzini** [comment]
Windows portability, and handling of shell quoting. But I guess that the first can be handled by the generator anyway, and the second doesn't matter for at least my use case. So if you think the patches are just too ugly to be worth it, feel free to close.

**Comment by bonzini** [comment]
Windows portability, and handling of shell quoting. But I guess that the first can be handled by the generator anyway, and the second doesn't matter for at least my use case. So if you think the patches are just too ugly to be worth it, feel free to close.
